### PR TITLE
[SPARK-18750][yarn] Avoid using "mapValues" when allocating containers.

### PR DESCRIPTION
**Pull request by vanzin**
That method is prone to stack overflows when the input map is really
large; instead, use plain "map". Also includes a unit test that was
tested and caused stack overflows without the fix.

### CODE REVIEW

**Comment by vanzin** [comment]
Argh, api not available in old hadoop... fix coming.

**Comment by vanzin** [comment]
@srowen @tgravescs

**Review by srowen — APPROVED** [review]
Looks fine. I'm surprised there is a difference but take your word for it if it fixes it. It should be functionally equivalent anyway.

File: resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/LocalityPlacementStrategySuite.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    val totalContainers = totalTasks / 16
+    val totalHosts = totalContainers / 16
+
+    val mockId = mock(classOf[ContainerId])
+    val hosts = (1 to totalHosts).map { i => (s"host_$i", totalTasks % i) }.toMap
+    val containers = (1 to totalContainers).map { i => mockId }
+    val count = containers.size / hosts.size / 2
+
+    val hostToContainerMap = new HashMap[String, Set[ContainerId]]()
+    hosts.keys.take(hosts.size / 2).zipWithIndex.foreach { case (host, i) =>
```
Comment (tgravescs) [inline resource-managers/yarn/src/test/scala/org/apache/spark/deploy/yarn/LocalityPlacementStrategySuite.scala]: similar wouldn't hurt to have small description here for someone looking at it later

**Review by tgravescs — COMMENTED** [review]
+1.  Just a couple small nits.

**Comment by tgravescs** [comment]
+1

**Pull request closed without merging**